enhancements/security/x509-trust.md

```diff
@@ -0,0 +1,174 @@
+This enhancement provides a framework for managing X.509 trust for these situations.
+
+Currently the default trust for CRI-O, the kubelet's cloud provider, and other node-namespace egress is the node's X.509 certificate authority (CA) store, which depends on the operating system ([Red Hat Enterprise Linux CoreOS (RHCOS)][rhcos], etc.) and any subsequent configuration ([the machine-config operator][machine-config-operator], etc.).
+The default trust for containers like the machine-API operator is that container's CA store, which depends on the base layer ([Red Hat Universal Base Image (UBI)][ubi], etc.).
+Many external components will have X.509 certificates signed by one of those default CAs, but this is not always the case.
+
+This enhancement provides a mechanism for configuring additional or alternative trust bundles for all cluster services when communicating with external components whose X.509 certificates are not signed by a default CA.
+
+## Motivation
+
+### Goals
+
+Administrators will have a single location to configure a default additional trust bundle to be used by all cluster components communicating with external components.
+They will also be able to specify whether this default is to be use in addition to, or instead of, the component's default trust bundle.
+
+Administrators will have per-component configuration locations to adjust the cluster-scoped default.
+As for the cluster-scoped default, they will be able to specify whether the per-component trust bundle is to be used in addition to, or instead of, the cluster-scoped default.
+
+### Non-Goals
+
+This enhancement does not address components making intra-cluster TLS requests.
+Trust for those requests is covered by [the Kubernetes API-server Certificates enhancement](../kube-apiserver/certificates) and [the Automated Service CA Rotation enhancement](../automated-service-ca-rotation.md).
+
+This enhancement does not address configuration for non-CA TLS settings like ciphers and allowed TLS versions.
+[The ingress TLS configuration enhancement](../kube-apiserver/tls-config.md) covers those settings for cluster-hosted services.
+Outgoing TLS probably needs similar configuration, but that configuration is decoupled from X.509 trust and should be addressed in a separate enhancement.
+
+## Proposal
+
+### Cluster-scoped Default Trust
+
+There will be a ConfigMap (FIXME: or should this be a new CRD under openshift/api?) named `default-ca-bundle` in the `openshift-config` namespace.
+The ConfigMap will have the following keys:
+
+* `ca-bundle.crt`, containing a PEM-encoded X.509 certificate bundle.
+* `composition`, configuring how the bundle in `ca-bundle.crt` is composed with the component's default trust store.
+    Valid values are:
+    * `append`, in which case components should use the union of the `ca-bundle.crt` and their default trust store.
+    * `override`, in which case components should use only `ca-bundle.crt` and ignore their default trust store.
```

Comment:
There are really two places to make this decision:

1) does the admin want to supply the total set of cluster CAs, or do they want to supply additionalCAs that should be appended to the cluster's system trust (as determined by the ubi image, basically)
2) does the each component accept the provided cluster CA bundle as the total set of CAs it trusts, or does it append the provided set to its own system trust bundle  (it will always append its own component-level CAs, if any)

We'd really only talked about (1) so far, but you seem to be talking about (2).

Reply:
> We'd really only talked about (1) so far, but you seem to be talking about (2).

This is basically "union with UBI defaults".  I don't really care if that happens Kube-side (1) or container-side (2).  I just assumed (2), because that's what we do now.  Do you have a reason to pivot to (1)?

Reply:
actually we do it somewhat kube side today and component side.

1) the proxy controller takes the additional CAs and appends the UBI system bundle to it
2) the proxy controller injects that new bundle into the labeled CMs
3) components consume that CM in whatever manner they choose.  They may use it as the only set of CAs, they might union it with their own system trusts, they might union it with their own other sources of CAs, etc.

Reply:
> actually we do it somewhat kube side today and component side.

Do you have a reason to preserve this, instead of pivoting to explicit `append` or `override` semantics?

> 1. the proxy controller takes the additional CAs and appends the UBI system bundle to it

Is this documented anywhere?  I don't see anything about it [here][1] or [here][2].

[1]: https://github.com/openshift/enhancements/blame/a16654eb55c5bcddd111cd59ed36b50d97a49385/enhancements/proxy/global-cluster-egress-proxy.md#L153-L154
[2]: https://github.com/openshift/api/blob/1271508fbe8fc691bd59c5dca8ad939ad7c9fded/config/v1/0000_03_config-operator_01_proxy.crd.yaml#L64-L75

Reply:
> Do you have a reason to preserve this, instead of pivoting to explicit append or override semantics?

we aren't going to (successfully) dictate what components choose to do with the content...  so i think on a component consumption perspective you will always have:
1) some that use what is provided as the only CAs
2) some that combine it w/ their own image system CAs
3) some that combine it with their own component-specific CA configuration

Plus (3) will probably always make sense for some components and not for others. meaning at least (1) and (3) are likely to always need to be allowed/implemented.  (1) vs (2) we can potentially declare a best practice rule about..I think we'd choose (1) there.


> Is this documented anywhere? I don't see anything about it here or here.

```
Today the additional CA bundle is a combination of user provided CAs plus the system CAs from the network operator image. In the future adding the system trusts to the bundle should be a configurable optional so customers who want to explicitly control the trusted CAs can do so.
```

Reply:
> we aren't going to (successfully) dictate what components choose to do with the content... 

We aren't?  If I gave the CVO a ConfigMap with my company CA and `override`, and someone MitMed my upstream with a cert signed by a sloppy/compromised big CA, and the CVO accepted the connection to the attacker, I would be really mad at the CVO and file a strongly worded bug ;).

Reply:
it's not just CVO, OLM operators will also consume this.  which will include ISV operators, community operators, etc.

Reply:
I expected a new trust bundle system to expand the existing additionalTrustedCA.  It’s not clear to me why we would have a second system?

https://github.com/openshift/api/blob/master/config/v1/types_proxy.go#L65

I expect to cover everything described here.  Why would we have a second one?

Or, if it’s not sufficient, how does something new interact?  How do they conflict?

Reply:
>  It’s not clear to me why we would have a second system?

Because you might have X.509 trust consumers that do not expect to be speaking to a proxy.  This enhancement spun off from openshift/installer#2658 where I proposed setting `TrustedCA` on the proxy config in the absence of [other proxy settings][1] to support the network-routed-proxy case, and the pushback was around the proxy config only being used for the [ask-apps-explicitly-to-use-a-proxy case][2].  And at least the CVO currently [only uses the proxy-config's `TrustedCA` when `HTTPSProxy` is set][3].  This proposal works around that pushback by defining a higher-level default trust config (the `default-ca-bundle` ConfigMap) that components should be clearly falling back to even if they feel like the proxy config doesn't apply to them.  So the CVO logic would be:

1. If `HTTPSProxy` is set (potentially also if `HTTPProxy` is set to an `https://` URI) and `TrustedCA` is set, use `TrustedCA` when validating TLS connections.  Otherwise...
2. If `default-ca-bundle` exists, use it when validating TLS connections.  Otherwise...
3. Use the container's default (UBI) trust store when validating TLS connections.

That would support use cases like:

> I'm not using a proxy, but I don't want my nodes trusting all the UBI CAs.  The CVO needs to talk to api.openshift.com (for Cincy, signed by Let's Encrypt) and mirror.openshift.com or storage.googleapis.com (for signatures, signed by DigiCert and Google Trust Services respectively).  Limit the CVO trust to those three CAs.

[1]: https://github.com/openshift/installer/pull/2658/files#diff-ff1d8a3b05383eef08ceaf043a235067R64
[2]: https://github.com/openshift/installer/pull/2658#issuecomment-553444380
[3]: https://github.com/openshift/cluster-version-operator/blob/b9a3ffafdc32e534c68112a13b89cf1459e1bbd9/pkg/cvo/availableupdates.go#L214-L219

Reply:
> potentially also if HTTPProxy is set to an https:// URI

Note that if `HTTPSProxy` is unset, `HTTPProxy` is used to proxy https requests regardless of the url scheme.